### ISSUE TITLE
Multiple ref in same line without whitespace fail to parse

### ISSUE THREAD

**Issue by rms1000watt** [variantdev/vals#13]
Template 1 in helmfile for logstash:
```bash
output {
  elasticsearch {
    hosts => ["ref+awsssm://pre0/pre1/pre2/?region=us-east-1#/host:ref+awsssm://pre0/pre1/pre2/?region=us-east-1#/port"]
  }
}
```

Error 1:
```bash
SSM: successfully retrieved key=/pre0/pre1/pre2/
in /Users/path/to/helmfile/stable/filebeat-logstash.yml: failed processing release logstash: expand awsssm://pre0/pre1/pre2/?region=us-east-1#/host:ref+awsssm://pre0/pre1/pre2/?region=us-east-1#/port"]: no value found for key host:ref+awsssm://pre0/pre1/pre2/?region=us-east-1#/port"]
```

It doesn't stop at the `:` which is a disallowed char

```Valid characters: Parameter names can consist of the following symbols and letters only: a-zA-Z0-9_.-/```

https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-parameter-name-constraints.html

I'll try to look at it too to see where this parsing happens, but I'm guessing you'll be able to find it faster 🤣 😊

**Comment by caseycs** [comment]
@rms1000watt inside `expand_match.go` you can find existing regex `((secret)?ref)\+([^\+:]*://.+)`, and it's greedy - so it's taking everything till the end of the line, defining a stop-characters might help you, for example `((secret)?ref)\+([^\+:]*://[\:]+)`, but it's a bit nasty

**Comment by caseycs** [comment]
Also my workaround might be useful for you: https://github.com/variantdev/vals/pull/16

**Comment by mumoshu** [comment]
@rms1000watt @caseycs Thanks for reporting!

Ah, this was something I didn't intend to support initially as written in [the doc](https://github.com/variantdev/vals#string-interpolation--template-functions), but I definitely need to change my mind.

To be clear, `vals` has been specifically designed assuming you'd have only one ref in a string. Now, we need to rethink it and have another syntax for embedding two or more refs in a single string.

The intention to use `ref+<provider-specific url>` was that it is expressive enough to represent any kind of single secret, while being capable of being embedded in any string without breaking validations. For example, you can safely [defer expanding refs until you run e.g. `kubectl apply`](https://github.com/variantdev/vals#helm). It would be if we could keep these points while adding support for multiple refs in a string.

#16 seems like a possible solution. The only concern atm is that I don't know if `#endref` doesn't collide with anything else.

Do we actually have other alternatives?

**Comment by rms1000watt** [comment]
An alternative would be a state machine for each type of ref and stop at the natural ending of each ref (since they're defined somewhere). I can't guarantee it's full proof for all refs, and it would be a smidge more work (although it's fun to implement a FSM).

Anyways, I'm ok with `#endref`. Can we consider `#end` to stick with Golang templating conventions of using `{{end}}` in their templating? Either way, I 👍!!

**Comment by mumoshu** [comment]
> An alternative would be a state machine for each type of ref and stop at the natural ending of each ref (since they're defined somewhere).

Reading this, a possible way came up to my mind was to let the user define the stop char, like ``((secret)?ref)(.)([^\+:]*://.+?)(\3)`. We need backreferences for this, but Go's regexp impl doesn't support it. So yeah we need a kind of parser to do it.

**Comment by mumoshu** [comment]
> let the user define the stop char at runtime

So you would use it like:


```
output {
  elasticsearch {
    hosts => ["ref|awsssm://pre0/pre1/pre2/?region=us-east-1#/host|:ref|awsssm://pre0/pre1/pre2/?region=us-east-1#/port|"]
  }
}
```

Or with another char:

```
output {
  elasticsearch {
    hosts => ["ref$awsssm://pre0/pre1/pre2/?region=us-east-1#/host$:ref$awsssm://pre0/pre1/pre2/?region=us-east-1#/port$"]
  }
}
```

**Comment by rms1000watt** [comment]
Whoa, that's super interesting, like `sed`.  👍 on any of these solutions, honestly.

The trick would be to make this change backwards compatible with the current setup. Not saying there are millions of people using this functionality already, but it would be a seamless feature addition.

**Comment by mumoshu** [comment]
Thanks!

If that seems to work, in short term, can we just start by changing the regexp pattern to have `+` as the stop char? It would be a simple change on top of #16 from @caseycs.

And later we can enhance it to allow defining arbitrary stop char(s) at runtime, which is still backward-compatible.

**Comment by rms1000watt** [comment]
👍 Sounds good to me!

**Comment by mumoshu** [comment]
@caseycs @rms1000watt Thanks a lot for your supports! I've implemented + as the stop char in c90d858. It seems to be working. But please let me know anything you find while using this in real world scenario.
